runtime/src/bank/builtins/core_bpf_migration/mod.rs

```diff
@@ -93,14 +93,38 @@ impl Bank {
     /// Create an `AccountSharedData` with data initialized to
     /// `UpgradeableLoaderState::ProgramData` populated with the current slot, as
     /// well as the source program data account's upgrade authority and ELF.
+    ///
+    /// This function accepts a provided upgrade authority address, which comes
+    /// from the migration configuration. If the provided upgrade authority
+    /// address is different from the source buffer account's upgrade authority
+    /// address, the migration will fail. If the provided upgrade authority
+    /// address is `None`, the migration will ignore the source buffer account's
+    /// upgrade authority and set the new program data account's upgrade
+    /// authority to `None`.
     fn new_target_program_data_account(
         &self,
         source: &SourceBuffer,
+        upgrade_authority_address: Option<Pubkey>,
     ) -> Result<AccountSharedData, CoreBpfMigrationError> {
         let buffer_metadata_size = UpgradeableLoaderState::size_of_buffer_metadata();
-        if let UpgradeableLoaderState::Buffer { authority_address } =
-            bincode::deserialize(&source.buffer_account.data()[..buffer_metadata_size])?
+        if let UpgradeableLoaderState::Buffer {
+            authority_address: buffer_authority,
+        } = bincode::deserialize(&source.buffer_account.data()[..buffer_metadata_size])?
         {
+            let target_upgrade_authority_address =
+                match (upgrade_authority_address, buffer_authority) {
+                    (Some(provided_authority), Some(buffer_authority)) => {
+                        if provided_authority != buffer_authority {
+                            return Err(CoreBpfMigrationError::UpgradeAuthorityMismatch(
+                                provided_authority,
+                                buffer_authority,
+                            ));
+                        }
+                        Some(provided_authority)
+                    }
+                    _ => None,
```

Comment:
I don't think `(_, None)` or can happen according to the loader rules, but I could also add a branch for this. In that case it would probably fail on `(Some(_), None)` and resolve to `None` on `(None, None)`.

Reply:
Yeah, even though we don't think buffer_authority.is_none() can ever happen, it would be nicer not to elide it in the catch-all case.
I think I would skip the match. (Suggestions in separate comment)